### ISSUE TITLE
Packer hangs while running shell provisioners with VirtualBox

### ISSUE THREAD

**Issue by Kounavi** [mitchellh/packer#2447]
Using VirtualBox 5.0 (& latest 4.x version) I've noticed that packer sometimes (1/10 from my experience) hangs after the first reboot. I'm always rebooting like this:

```
reboot now
sleep 60
```

per the recommended practices I've seen in other posts so ssh may be killed reliably.

Is anyone aware of this behaviour?

**Comment by cbednarski** [comment]
@Kounavi there are a lot of things that could be happening during a hang. Here are a few things you can try:

- Enable packer logging user `PACKER_LOG=1 packer build <template>`. This should show more verbose debug and error messaging so you can see exactly where the hang occurs and what packer is doing.
- Use `shutdown -r now` instead of `-h`? `-h` tells the box to remain powered off after shutdown. `-r` tells it to restart instead.
- Activate the GUI so you can see what's going on. Perhaps the box is never able to shutdown or never comes back up. I think you'll want to use [`headless=false`](https://www.packer.io/docs/builders/virtualbox-iso.html#headless).

**Comment by Kounavi** [comment]
Ok, I've come up with more info.
I'm able to reliably now (kind of) replicate this. Here's the scenario:
I download some GPG signatures, verify repositories against them and then update the OS, download some complimentary packages and finally download some gui packages before rebooting.
The first time this happened I logged in to the server ( flag 'headless=false' always when developing ;) ) it stuck there as you can see in the screenshot. However, the first stage before reboot had completed.
The second time, I re-run ``packer build <filename>.json`` and it stuck right after the first GPG signature download (with all messages printed until there in contrast to the first step where everything had finished before reboot - log files are the proof of it :) - but with no message output in the terminal.

Considering that:
My YUM repository is replicated and I use a local FTP repository instead combined with the fact that I am over a public, **crappy**, connection during the daylight has led me to think that poor connections cause the ssh daemon or packer to hang up. I'll try next time and use PACKER_LOG=1 so I may delve into this further. ;)

P.S.: I updated my answer. I had erroneously pointed out that I was doing a ``shutdown -h now`` but instead I use ``reboot now``. I use ``shutdown -h now`` only in the last step (and have no problem with that). Finally, I use packer  v 0.7.5 so I'll try tomorrow with the new and see how it goes for a few days :)

![packer_vbox_ssh_hang_instance](https://cloud.githubusercontent.com/assets/540404/8738179/2328fedc-2c34-11e5-8cf8-f8378321e879.jpg)

**Comment by cbednarski** [comment]
Thanks for digging into this and providing more information.

Packer uses retries and timeouts to try to talk to the VM, even if things are taking awhile. When you run with `PACKER_LOG=1` you should see these retry loops happening.

We identified a few places where timeouts were missing. For example, we recently fixed a timeout that happened during the SSH handshake when trying to login to the VM. There may be others that we have missed. `PACKER_LOG=1` should help identify where it's getting stuck.

It's also possible that the network problem is caused by / happening inside the VM. In that case packer should eventually timeout and abort, though I recall some of the default timeouts are quite long (5 minutes or more) so if you're watching the terminal it may seem to you like packer is stuck.

**Comment by Kounavi** [comment]
No problem! :) Always happy to help ditching out problems. It;s for the benefit of everyone - and you! :)
One last detail: Is there any way to redirect ``PACKER_LOG=1`` output to a file and keeping the echo statements in terminal? I don't like the clutter and all this stuff since it makes it hard to debug even my own code :)

**Comment by cbednarski** [comment]
> Is there any way to redirect PACKER_LOG=1 output to a file and keeping the echo statements in terminal?

Sorry I think it's all or nothing. Using the `-debug` flag might write the log to a file under `/tmp` but that changes execution in other ways too.

**Comment by cbednarski** [comment]
@Kounavi Turns out you can use `PACKER_LOG_PATH` to send the logs somewhere else, and they won't clutter your terminal. Found this on https://packer.io/docs/other/debugging.html Cheers!

**Comment by Kounavi** [comment]
All right! I'l let you know soon since I do notice some problems even when using Packer 1.8 :)

**Comment by Kounavi** [comment]
Oh no! I don't know but the newer version gets stuck all the time somewhere :(
Let me be straight: 
First, I consistently had messages of this type:
![packer_0](https://cloud.githubusercontent.com/assets/540404/8889846/cb1104bc-32f3-11e5-8df4-736334ff7bfa.png)
Then, another building process that hangs in the middle:
![packer_1](https://cloud.githubusercontent.com/assets/540404/8889849/2236c880-32f4-11e5-8580-ab57c84f22b3.png)
Since I saw a little bit late your message about debug - very sorry - time to make use of it:
![packer_2](https://cloud.githubusercontent.com/assets/540404/8889855/4beac334-32f4-11e5-9406-4f52b17daae3.png)
I'm attaching the log. I've noticed that after the output stuck there the log was populated and everything seemed to work ok (I assume there's buffering or a routine/function that keeps system output at hand for sometime before flushing it?). Until the log prints this:

https://gist.github.com/Kounavi/2d74272cbbe19937e8ba#file-packer_log_end-txt

What I understand so far: After the message from my script 'Installing guest additions' is printed packer reboots where I expect the process to complete as it always does by printing successful messages. Why? :( In addition, I think this is a race condition since it reboots out of the blue and goes on with the next step of the provisioning process. After that, it does nothing!
Something's wrong apparently. I've logged to the server and guest additions where installed so why did packer didn't print the messages? If it had failed (based on the log output) then the guest additions wouldn't have been installed. However:

I installed pciutils and lspci shows this:
00:02.0 VGA compatible controller: InnoTek Systemberatung GmbH VirtualBox Graphics Adapter
Also, in /etc/inittab the entry ``id:5:initdefault:`` exists as expected which means my script modified it after guest addition installationation since it was set ``id:3:initdefault:``.
Really confused here and I'm facing various so I'll probably stick to an older packer variant because this one creates me a lot of unpredictable hangs :(

P.S: Log is available here: https://gist.github.com/Kounavi/2d74272cbbe19937e8ba

Do you know where may I download previous packer version (i.e. 0.7.5)?

**Comment by cbednarski** [comment]
Thanks for following up! If you want to download a previous version you can change the version number in the download URLs on this page: https://packer.io/downloads.html For example windows 64-bit would be:

https://dl.bintray.com/mitchellh/packer/packer_0.7.5_windows_amd64.zip
https://dl.bintray.com/mitchellh/packer/packer_0.8.2_windows_amd64.zip

etc.

**Comment by Kounavi** [comment]
Ok. Some more info here. I am almost consistently facing retryable ssh errors when using powershell for building. 9/10 times it gets stuck at the last step where it shutdowns. However, this time it got stuck where it says 'Installing OpenStack Glance service'. Now, it seemingly hanged in there but the machine performed at least one reboot (maybe more). So, intuitively I pressed **once** Ctrl-C and the output got unstuck! Then it finished and exported the image. Maybe a buffering issue?

P.S.: No debug logs. The reason is that I've environmental variables are ignored (packer_log & packer_log_path) and the cmd is even more unreliable when building (it gets stuck randomly with surgical precision at some step). I'll try and workaround the powershell issue otherwise I will try the cmd since I understand that output logs are useful for debugging, reverse engineering, fixing bugs etc :)

![packer_success_with_ctrl_c](https://cloud.githubusercontent.com/assets/540404/8900891/ba14038a-344c-11e5-9d1d-0b032c65943e.png)

P.S.: Sorry for the long, detailed posts. I wish they were shorter and more concrete :(

**Comment by cbednarski** [comment]
This is definitely weird behavior but there are a lot of things that are potentially going wrong. Can you share a minimal packer template that produces this problem?

If you're using winrm this might be caused by a [race condition in the winrm code](https://github.com/mitchellh/packer/issues/2312). It's also possible that this is related to another issue where [shell output is not fully received from the remote side](https://github.com/mitchellh/packer/issues/2391).

**Comment by Kounavi** [comment]
I have posted a gist link with various logs. If you are talking about the latter specific error about the output then I still don't have logs for that. Currently, my approach is for before & after every reboot I let the script sleep for 60 seconds. Maybe this helps packer catch-up with the progress of the script file. No, I'm not using winrm. Yes, I can share a packer template but I can't share the scripts that the template uses since it's proprietary code :( (maybe share a private gist with you?)

**Comment by cbednarski** [comment]
> P.S.2: Shouldn't the https://www.packer.io/docs/other/debugging.html be updated so as to demonstrate how to do that with Powershell

Yes that would be great!

> P.S.3: A debug log has been posted here:

Awesome, very helpful. Thanks for providing all this info. I will see if I can diagnose this for the patch following 0.8.3.

**Comment by Kounavi** [comment]
I've uploaded another packer log. I don't know if it's useful but it is deterministic:
https://gist.github.com/Kounavi/6cdfb35dac4dc38d0fe1#file-packer_retryable_error-txt
This one however causes a retryable error due to SSH handshake.
Packer version is 0.8.5

**Comment by Kounavi** [comment]
Sharing the JSON files is ok (they are autogenerated). If you would like access to them let me know (the script however is another story :P).

**Comment by Kounavi** [comment]
Here's a a successful build after waiting so long:
https://gist.github.com/Kounavi/6b14e6e8dc62acbec614

This one however had its ssh_timeout_wait increased to 50m since 45m caused an ssh retryable error in a consistent and deterministic way. 

![packer_network](https://cloud.githubusercontent.com/assets/540404/9941854/9b286bc0-5d80-11e5-8885-0ab4fdadda6d.png)

At 4:00 am packer started the build process.
At 4:44 am it was stuck there for like 2 hours.
At 6:44 am it was waiting there for some time and after some time the build was completed successfully.
We're talking here about 3 hours of build time.
So unproductive :(

**Comment by Kounavi** [comment]
Is there any update about this bug? The latest version (0.9.0) now hangs when '-debug' is used :(
Not trying to be prick or pushy here but these random hangs are making me totally unproductive and there is no workaround. Any ideas of whether you aim to incorporate some mechanism to force packer resume its operations?

**Comment by rickard-von-essen** [comment]
@Kounavi There is way to much noise here. Could you provide:
- A minimal config to reproduce this. Gist a template and your scripts
- Gist a log (```PACKER_LOG=1 packer build ...```) where you run that and get the error.
- Some details on versions (```VBoxManage --version```, OS version etc.)

**Comment by cbednarski** [comment]
I ran across this today using 0.10.0. The VM was responsive -- I could login via the GUI -- but provisioners stopped running for some time.

I looked at the code changed between 0.7.5 and current master and there are several places where there have been significant changes in VirtualBox:

1. SkipNatMapping was added
2. VRDP was added
3. VMware/ Virtualbox now use consolidated communicators

All of these are network-related changes, though this may be a red herring.

In the hang I observed it would frequently be related to a shell script that called out to the network, and would get stuck between individual commands in the same script, so it appears that either packer is no longer able to communicate with the VM or the command is blocking in the VM, possibly waiting on network IO.

Since the log posted above shows a rather specific amount of time (2h) after which the script resumes, I'm curious whether there is a deadlock somewhere in the communicator code where one of the goroutines eventually times out and the provisioner is able to continue. 2h seems way too long to be a network timeout.

At this point I can also not rule out the possibility that this is a bug in VirtualBox but I think there are a few places this could be caused by Packer so I will try to dig more here before blaming the VBox team. :)

> The latest version (0.9.0) now hangs when '-debug' is used :(

This is now fixed on master. I don't think these issues are related but I will try to verify.

Also, I ran a slew of VMware builds during the same time period and I never encountered this problem with VMware, so definitely seems to be something specific to VirtualBox.

**Comment by cbednarski** [comment]
This code path is not covered by tests. I'd like to add a provisioner test to https://github.com/mitchellh/packer/blob/master/builder/virtualbox/iso/builder_acc_test.go and see if `go test -race` picks anything up.

**Comment by cbednarski** [comment]
I'm wondering if this is the same issue: https://github.com/mitchellh/packer/issues/507

**Comment by heckj** [comment]
potentially related - while we were hunting down a similar sounding bug, we found that if a provisioner script started some services using `upstart` with an Ubuntu base image and left it running, we could reproduce a hang on some systems.

explicitly turning off those services before completing the individual shell script resolved this issue, so I'm presuming it held some SSH session detail open that would otherwise close naturally with the existing code paths.

Our bug related to this (if anyone's interested): https://github.com/RackHD/RackHD/issues/372, and the resolution in updating the provisioning scripts: https://github.com/RackHD/RackHD/pull/374

**Comment by mwhooker** [comment]
We just updated our ssh library. I wonder if that would have any impact on this issue https://github.com/mitchellh/packer/pull/4546

**Comment by mwhooker** [comment]
We've done a lot of work around this part of the code, and so I wonder if it's still an issue. This ticket has been opened a while so a lot of the information is quite stale. I'm going to close this since I can't confirm that it's still an issue and it's hard to know what the "fix condition" is. If someone is still running into this with a recent build of packer, I encourage them to open a new issue. I would hate for someone to be getting failed builds because of this, so I really want to fix it, but I don't see anything actionable in this ticket right now. Thanks!

**Comment by unixzen** [comment]
New similar issue 

https://github.com/hashicorp/packer/issues/6308
